library/std/src/sys/pal/uefi/helpers.rs

```diff
@@ -142,7 +147,48 @@ pub(crate) unsafe fn close_event(evt: NonNull<crate::ffi::c_void>) -> io::Result
 
 /// Get the Protocol for current system handle.
 /// Note: Some protocols need to be manually freed. It is the callers responsibility to do so.
-pub(crate) fn image_handle_protocol<T>(protocol_guid: Guid) -> Option<NonNull<T>> {
-    let system_handle = uefi::env::try_image_handle()?;
-    open_protocol(system_handle, protocol_guid).ok()
+pub(crate) fn image_handle_protocol<T>(protocol_guid: Guid) -> io::Result<NonNull<T>> {
+    let system_handle = uefi::env::try_image_handle().ok_or(io::const_io_error!(
+        io::ErrorKind::NotFound,
+        "Protocol not found in Image handle"
+    ))?;
+    open_protocol(system_handle, protocol_guid)
+}
+
+pub(crate) fn device_path_to_text(path: *mut device_path::Protocol) -> io::Result<OsString> {
+    let device_path_to_text_handles = locate_handles(device_path_to_text::PROTOCOL_GUID)?;
+    for handle in device_path_to_text_handles {
+        let Ok(protocol) = open_protocol::<device_path_to_text::Protocol>(
+            handle,
+            device_path_to_text::PROTOCOL_GUID,
+        ) else {
+            continue;
+        };
+
+        let path_ptr: *mut r_efi::efi::Char16 = unsafe {
+            ((*protocol.as_ptr()).convert_device_path_to_text)(
+                path,
+                r_efi::efi::Boolean::FALSE,
+                r_efi::efi::Boolean::FALSE,
+            )
+        };
+
+        // SAFETY: Error only occurs if path is NULL. Thus no need to dealloc path_ptr
+        let path_units = unsafe {
+            WStrUnits::new(path_ptr)
+                .ok_or(io::const_io_error!(io::ErrorKind::InvalidData, "Invalid path"))
+        }?;
+
+        let path_vec = path_units.map(Into::into).collect::<Vec<u16>>();
```

Comment:
I think this allocation can be eliminated as we can get a `&[u16]` from the pointer returned by UEFI, we only need to compute the length here:
```suggestion
        let path_len = path_units.count();
        let path_slice = unsafe { slice::from_raw_parts(path_ptr.cast(), path_len) };
        let path = OsString::from_wide(&path_slice);
        // etc...
```